security/security-secrets-setup.1.rst

```diff
@@ -0,0 +1,191 @@
+NAME
+====
+
+security-secrets-setup — Creates an on-device public-key infrastructure (PKI) to secure microservice secret management
+
+SYNOPSIS
+========
+
+| security-secrets-setup [-confdir confdir] -config params.json
```

Comment:
@anonymouse64 Ian, I guess we have misunderstanding here: every flag here is a choice and hence option. For example, you cannot run generate with other option cache or import. plus, there is no command to start with if using sub-command. `security-secrets-setup` to me is like an CLI executeable not command. and thus we do not have command to start with.  As the end result,  `generate`, `cache`, or `import` becomes option or choice flag.   Is that making sense to you?

Reply:
Unfortunately I don't quite follow your line of reasoning...

> every flag here is a choice and hence option

Sure, but I'm not trying to philosophically argue that subcommands are not a user's choice, just that it makes more sense for an option to be optional and a subcommand to be required than the other way around. To put it more bluntly, options should be optional.

> there is no command to start with if using sub-command

That's fine, I have on my machine dozens of applications that don't do anything without a sub-command which I use frequently. What most of them do is to just print a help message when run without a sub-command.

> security-secrets-setup to me is like an CLI executeable not command

I don't follow what the distinction is between a CLI executable and a command?

Reply:
Well, one thing it looks really strange to me is that using subcommand without any options or choice associated with it because *usually* you choose to implement `subcommand` approach due to multiple choices to associate with it.  For example, `git commit -m "xxx" -s -a` to support `-m` option, `-s` option, and `-a` option associated with `commit` subcommand.  However, in the current case of `generate`, `cache`, and `import` are just one function by itself and each one individually mutually exclusive.  So I do not have the strong use cases or reason to implement it as subcommand or `flagset` in go's terminology.  But I can see one plus side to use subcommand is extensibility, though it introduces complexity that we currently do not need I think.

Reply:
Again, I think `git` is a great example of using subcommands correctly here instead of options. Note that the `-m` and `-s` options are **_optional_**, what is necessary is `commit` which instructs `git` what kind of thing is to be done, the options are instructing how to modify what is being done, but note that you can just run `git commit` by itself without any options and it still works, you get an interactive editor with default commit options. 

What you're proposing here is that the options, `-generate`, `-cache` and `-import` are not **_optional_** since you need at least one of them, so since you need at least one of them the usual way that's implemented is with a subcommand. 

Some other examples of applications that work like this:
- `vault`
- `consul`
- `snap`
- `lxd`
- `docker`

So we can try to get some resolution here, @jim-wang-intel can you give some emperical examples of command line applications which work the way you're proposing `security-secrets-setup` work with these non-optional options?

Reply:
I agree sub commands look fancier, but with the promise of Fuji release deadline coming to so close, I am inclined to keep what we have currently UNLESS it is a major bug or "must be" fix. @anonymouse64 I would suggest to open an issue to track the suggestion of having sub command so that we can address it when we have resource available.

Reply:
> So we can try to get some resolution here, @jim-wang-intel can you give some emperical examples of command line applications which work the way you're proposing `security-secrets-setup` work with these non-optional options?

@anonymouse64 Ian, lots of linux commands have many examples of just using options.  For example, `ls`, `netstat`,...etc.   They are implemented with dash - options.

Reply:
But neither of those commands have **_non-optional options_** (this is the biggest point I keep trying to make), both `netstat` and `ls` work without any options because they only do one thing and their options only slightly change what they are doing

Reply:
All have agreed on conference call that we will use a subcommand. Pushed https://github.com/edgexfoundry/edgex-docs/commit/3f698b30a4121bdfc849cc02229a644a27c5e6e2.